spring-cloud-gcp-security-iap/src/main/java/org/springframework/cloud/gcp/security/iap/AudienceValidator.java

```diff
@@ -44,19 +47,10 @@
 
 	private String audience;
 
-	public AudienceValidator(String audience) {
-		Assert.notNull(audience, "Audience cannot be null");
-		this.audience = audience;
-	}
-
-	/**
-	 * Allows subclasses to delay setting the audience value after the object is constructed.
-	 */
-	protected AudienceValidator() { }
-
-	protected final void setAudience(String audience) {
-		Assert.notNull(audience, "Audience cannot be null");
-		this.audience = audience;
+	public AudienceValidator(Supplier<String> audienceSupplier) {
+		Assert.notNull(audienceSupplier, "Audience cannot be null");
+		this.audience = audienceSupplier.get();
```

Comment:
I don't think it is a good idea to evaluate supplier exactly here in the ctor.
On the other hand if we still do this, what is the point then request a supplier wrapper if we could just ask end-users to inject the target `audience` result?

If we still have some hard logic about  `audience` evaluation, I would suggest to introduce a new `AudienceProvider` interface instead of using `Supplier` to not confuse Spring Cloud Function (see my comment above), but still don't evaluate the target result here in the ctor: some low level resources interactions may present in the provider implementation.

Reply:
aaaand I should have read all comments first. Thanks!

Reply:
Hmm, now that I think about it, it seems better to only evaluate the audience string in the constructor and set it forever. The constructor runs once, but JWT token validation could happen at high volume. It seems wasteful to keep re-request information about the project and keep concatenating what is essentially a static string over the lifetime of the app.

An alternative would be to cache the string in each AudienceProvider, so that it's only evaluated once.

Reply:
You can cache it in this `AudienceValidator` implementing the evaluation logic in the `afterPropertiesSet()`. See `InitializingBean`.